libhb: make source audio track name available to frontends

jstebbins opened the pull request:
It gets stored in new hb_audio_config_t.in.name field.

It is available in the title dict read through hb_title_to_dict() or
hb_title_to_json() in AudioList[].Name.

When a job is initialized with hb_preset_job_init or
hb_preset_job_init_json(), output audio tracks are initialized with the
source track name.

Also adds output track name initialization to LinGui.

**Test on:**

- [x] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [x] Ubuntu Linux

jstebbins committed:
libhb: make source audio track name available to frontends

It gets stored in new hb_audio_config_t.in.name field.

It is available in the title dict read through hb_title_to_dict() or
hb_title_to_json() in AudioList[].Name.

When a job is initialized with hb_preset_job_init or
hb_preset_job_init_json(), output audio tracks are initialized with the
source track name.

Also adds output track name initialization to LinGui.

sr55 commented:
I'm planning to just pass this through if it's not null. Does anyone have any objections?  @bradleysepos  @galad87

bradleysepos commented:
Seems that would clobber a lot of inputs. Is it feasible to keep the existing behavior where null or make this a preference?

sr55 commented:
what do you mean?  The existing behaviour is a loss of track metadata. So it defaults to "Stereo" or whatever we set it to internally.

bradleysepos commented:
Nevermind, pass through if *not* null is what we want. I thought you were passing through null. We should do something smarter in that case if the name is set but empty/null.

jstebbins commented:
There hasn't been any libhb default for output track names.  It's empty.  LinGui has always just left it empty with the option to set it manually in the audio track settings.

There is a preset key "AudioTrackName" that gets applied in hb_preset_job_init if it's set, but no builtin presets use this key and LinGui has no option to set this key.  With this patch, the key is used first, and if not set the source track name is used in hb_preset_job_init.

sr55 commented:
Maybe being daft but I don't think this is working:

![image](https://user-images.githubusercontent.com/628593/62247846-9ecf2d00-b3de-11e9-8f4a-a0f7c2cf4e0d.png)

![image](https://user-images.githubusercontent.com/628593/62247855-a4c50e00-b3de-11e9-9e43-85941b4b3ea0.png)

sr55 commented:
Might be my fault. checking

jstebbins commented:
Are you using a source that has an audio track name?  I had to create one.

sr55 commented:
The media info is one i created, with "Test" as the name. (created by us)

Just recompiled and nada :(

Also, what's CodecParam?

jstebbins commented:
This is what I get from hb_title_to_json with my sample
```
{
    "AngleCount": 1,
    "AudioList": [
        {
            "Attributes": {
                "AltCommentary": false,
                "Commentary": false,
                "Default": true,
                "Normal": false,
                "Secondary": false,
                "VisuallyImpaired": false
            },
            "BitRate": 640000,
            "ChannelCount": 6,
            "ChannelLayout": 1551,
            "ChannelLayoutName": "5.1(side)",
            "Codec": 2048,
            "CodecName": "ac3",
            "CodecParam": 86019,
            "Description": "English (AC3) (5.1 ch) (640 kbps)",
            "Language": "English",
            "LanguageCode": "eng",
            "Name": "My Audio Name Test",
            "SampleRate": 48000
        }
    ],
    "ChapterList": [
        {
            "Duration": {
                "Hours": 0,
                "Minutes": 0,
                "Seconds": 8,
                "Ticks": 720000
            },
            "Name": "Chapter 1"
        }
    ],
    "Color": {
        "Format": 0,
        "Matrix": 1,
        "Primary": 1,
        "Range": 1,
        "Transfer": 1
    },
    "Container": "matroska,webm",
    "Crop": [
        0,
        0,
        0,
        0
    ],
    "Duration": {
        "Hours": 0,
        "Minutes": 0,
        "Seconds": 8,
        "Ticks": 720000
    },
    "FrameRate": {
        "Den": 1125000,
        "Num": 27000000
    },
    "Geometry": {
        "Height": 544,
        "PAR": {
            "Den": 1,
            "Num": 1
        },
        "Width": 1280
    },
    "Index": 1,
    "InterlaceDetected": false,
    "Metadata": {
        "Name": "Sintel.2010.720p"
    },
    "Name": "Sintel.2010.720p",
    "Path": "/home/jstebbins/Videos/Sintel.2010.720p.mkv",
    "Playlist": -1,
    "SubtitleList": [],
    "Type": 3,
    "VideoCodec": "h264"
}
```

sr55 commented:
Works for a MKV file, doesn't work for an MP4 file

jstebbins commented:
do you have MP4Box handy?  Can you run MP4Box -diso on the file and find what atom the title is in?  My be ffmpeg doesn't parse it.

sr55 commented:
[Namedtrack2_info.txt](https://github.com/HandBrake/HandBrake/files/3453961/Namedtrack2_info.txt)

Attached output

sr55 commented:
Windows UI is ready to go.

jstebbins commented:
So, libavformat writes tag "name" when "title" track metadata is given, but reads tag 0xa9"nam" into "title" track metdata.  It doesn't read and write the same tag :*(

jstebbins commented:
ffmpeg is all kinds of broken.  it doesn't process that tag, it doesn't process the apple spec'd tnam tag.  and in fact it assumes all udta tags are global, not track specific.  I'll see if I can figure out a patch, but i'll have to detect that a track is being processed somehow rather than a higher level udta atom.

sr55 commented:
Sorry, I opened a can of worms again :(

jstebbins commented:
I found a solution.  Testing patch...

jstebbins committed:
ffmpeg: add 'name' track udta tag reading

jstebbins commented:
Fixed

galad87 commented:
The correct way to make Apple things read the title is to use the "titl" atom. There was a bug where they wrote "tnam" instead of "titl" so they are continuing to support "tnam" too.

jstebbins commented:
Old qtff documentation only lists tnam
https://developer.apple.com/library/archive/documentation/QuickTime/QTFF/QTFFChap2/qtff2.html

Do you have a source for documentation on titl?

galad87 commented:
It's in the 3gpp specs: https://www.etsi.org/deliver/etsi_ts/126200_126299/126244/06.02.00_60/ts_126244v060200p.pdf

galad87 commented:
Apple uses "tnam" in mov, but "titl" in iso file format / mp4.

jstebbins commented:
ffmpeg will write 'titl' tag in a udta atom for the *global* title only when the "mode" is 3gp.  For mp4 "mode" it uses '©nam' tag in an meta.mdir atom rather than udta (main global title).  For tracks, it writes 'name' tag in udta in mp4 and mov "mode" and nothing in other modes. The format it writes this in doesn't match any documented format I can find.

Without my patch, it doesn't read any track metadata.

The latest qtff document I have (2012) documents meta.mdta for metadata and has a list of "QuickTime defined keys" of the format "com.apple.quicktime.title". Keys in mdta can be of any length.  

ffmpeg writes meta.mdir style metadata by default (for global meta), but has a flag to enable writing meta.mdta style.  It uses the same tag values either way.

It will write 3gpp udta metadata only in 3gp mode (and only global tags). It has a separate list of tags for this mode.

I can't find any documentation on mdir, I don't know who invented that.

Thanks for the 3gpp link.  

I guess the summary is that tags in mp4 are a mess with multiple specifications and zero agreement on what keys to use.

jstebbins commented:
and it appears ffmpeg doesn't read 3gpp format udta tags at all.

sr55 commented:
Dare I ask the state of play for naming subtitle tracks also?

jstebbins commented:
Wouldn't be all that hard to add.  Similar amount of code changes in libhb as audio was. I don't know about the other GUIs but I never added an entry field to the subtitle window for a name.  So I'll have to add that. 

Another thing we might want to consider is constructing names from the descriptive attributes available for streams on DVD (e.g. Director's Commentary).  Unfortunately, BD doesn't have equivalent descriptors, so not much we can do there.

galad87 commented:
@jstebbins Apple uses only the "titl" atom from 3gpp specs for the track name. All the globals metadata are in the iTunes format.

In Subler I write both "titl" and "name" for track titles. It works ok on all the Apple devices that can display track names.

jstebbins commented:
What atom format do you use for "name"?  ffmpeg is using a "raw" format that doesn't conform to any of the specifications.

Which "itunes" format for globals?  The most recently documented format I've found is udta.meta.hdlr with handler type "mdta" in the qtff 2014 spec.  The "itunes" format ffmpeg writes by default is udta.meta.hdlr with handler type "mdir". My google fu fails to find anything other than more ffmpeg references for mdir. ffmpeg has a flag that can be set to output the mdta variant.

I have several versions of qtff docs (2001, 2007, 2012, 2014). The first version I have with the mdta variant is 2012.  In the 2007 version they don't document 'meta' atom at all.

jstebbins committed:
a few fixes for audio track name handling

fixes a memory leak makes name const

jstebbins committed:
add subtitle track name read/write

Works similar to audio track names.
If source has a subtitle track name, hb_subtitle_t.name is set.
To set output subtitle track name, set hb_subtitle_config_t.name.

Source track names are available in title returned by hb_title_to_dict
and hb_title_to_json in SubtitleList[].Name

In job dict it is also SubtitleList[].Name

hb_preset_job_init and hb_preset_job_init_json initialize output tracks
with the source track name.

Also adds subtitle name support to LinGui

jstebbins committed:
add subtitle name to job configuration log output

jstebbins committed:
CLI: add subtitle name support

jstebbins commented:
> Dare I ask the state of play for naming subtitle tracks also?

There you go :+1:

jstebbins committed:
muxavformat: fix writing alternative track name in mp4 hdlr atom

ffmpeg changed the name of the metadata key at some point

jstebbins committed:
muxavformat: set handler_name for subtitle track names also

galad87 commented:
iTunes metadata are something like this: http://atomicparsley.sourceforge.net/mpeg-4files.html
But it's unrelated to track names.

Here's a sample with both iTunes global metadata and both tracks name: https://subler.org/downloads/test.mp4

sr55 commented:
Looks like it doesn't compile.

../libhb/muxavformat.c: In function 'avformatInit':
../libhb/muxavformat.c:972:67: error: 'name' undeclared (first use in this function); did you mean 'nanl'?
                 av_dict_set(&track->st->metadata, "handler_name", name, 0);
                                                                   ^~~~
                                                                   nanl
../libhb/muxavformat.c:972:67: note: each undeclared identifier is reported only once for each function it appears in
../libhb/module.rules:12: recipe for target 'libhb/muxavformat.o' failed
make: *** [libhb/muxavformat.o] Error 1
make: *** Waiting for unfinished jobs....
In function 'hb_strncat_dup',
    inlined from 'hb_autopassthru_print_settings' at ../libhb/common.c:2586:23:
../libhb/common.c:5388:9: warning: 'strncat' specified bound 2 equals source length [-Wstringop-overflow=]
         strncat( str, s2, n );
         ^~~~~~~~~~~~~~~~~~~~~

jstebbins committed:
fix build error

not sure how/why I committed somethng I didn't even compile :-\

jstebbins commented:
@galad87 that page documents everything *but* the hdlr atom where the difference exists between what is documented in qtff specs and what we see in the wild.  But no worries.  That's what ffmpeg outputs by default for global tags.  The 'name' tag matches the raw tag output ffmpeg does. MP4Box doesn't recognize the titl tags so is outputting just generic xml description.  So I looked at it with a hex editor and it is indeed the 3gpp format udta tag.

So basically, apple uses 3 different formats for tags, 2 of which have no official documentation.

Weird, but whatever :man_shrugging: Do you know of any software or devices that recognize titl and not name?  ffmpeg doesn't currently support writing titl, but I could certainly patch it easily enough.

@sr55 sorry, not sure how I managed to commit without even compiling.

galad87 commented:
I think you should just ignore the qtff specs.
So there are only two type of metadata: iTunes style metadata for global metadata (artist, release date, track number, etc), and 3gpp "titl" for the track title.

Apple softwares recognise only "titl" and not "name" for track titles.

jstebbins commented:
Ah, thanks for the clarification.

It's funny, I'm the one that added the code to ffmpeg to write the 'name' tag.  So there must have been some software that was using it at the time.  I took another look at the qtff spec and 'name' *is* documented there.  I was just confused by their discussion about language codes.  Some of the tags in the table require language codes in the atom and some don't.  'name' does not.

Anyway, I'll add a patch to also write the 3gpp style titl tag.  To push upstream, it may require also adding an AVOption flag to enable it.  They may not want to change the default output.

galad87 commented:
Yes Apple used to support "name" in QuickTime Player 7, but when they rewrote everything from scratch in the new AVFoundation framework they choose to use only "titl". ¯\_(ツ)_/¯

bradleysepos commented:
So, if we're supplying a track name from scratch, we write both. If we're passing through, we pass through and make sure both are populated. But what do we do if the original has both and they differ? Pass through as-is?

sr55 commented:
**Subtitles**

Writing a Track Works:
![image](https://user-images.githubusercontent.com/628593/62415882-f9f46000-b628-11e9-9166-c831a7a2f889.png)

![image](https://user-images.githubusercontent.com/628593/62415894-29a36800-b629-11e9-8b20-6a275d6501df.png)

Reading back the Track Works:
![image](https://user-images.githubusercontent.com/628593/62415889-185a5b80-b629-11e9-8801-14528c89a23c.png)

**Audio**
Appears to already be getting automatic naming if left blank. I.e. "Stereo".

Otherwise:
MP4:

![image](https://user-images.githubusercontent.com/628593/62415977-8f442400-b62a-11e9-9068-251ccb828361.png)

Write: Working
Read: Not Working Yet

MKV:

![image](https://user-images.githubusercontent.com/628593/62415980-9ec36d00-b62a-11e9-9088-127b84b07af0.png)

Write: Working
Read: Working

sr55 commented:
Linking issue: #855

sr55 commented:
Windows GUI updates are ready to go one y'al have done fussing over MP4 :)

jstebbins commented:
The current automatic naming for audio is based only on channel layout (mono, stereo, multi-channel) and happens in muxavformat.  We could do something smarter with DVDs where there are attributes describing the type of track (e.g. commentary, children, visually impaired).

> Read: Not Working Yet

What is not working.  For me, we can read everything that we write.  So what is the test case that fails?  Maybe you didn't rebuild ffmpeg after I added the patch to read the 'name' tag?

sr55 commented:
It Failed to read a file HB wrote with MP4. I'll re-check later to make sure I didn't screw something up.

jstebbins committed:
ffmpeg: add support to read/write mp4 'titl' track tag

jstebbins commented:
@galad87 hopefully that last commit makes the track tags apple compatible

sr55 commented:
All Good my end.  ffmpeg hadn't recompiled after the patches added.

Happy with this when everyone else is

galad87 commented:
@jstebbins almost. If you set the title language to English it will only show it if the system language is set to English too.

jstebbins commented:
Ok.  I'm at a conference, so it'll be a few days before I can work on this more.

jstebbins committed:
stream: improve track name selection

ffmpeg will suppliy mulitple metadata entries for 3gpp track name
entries if there are language specific names in the source.  Pick the
"best" option.

jstebbins committed:
ffmpeg: make track names visible regarless of locale

Apple will not show a track name if it doesn't match the current locale
or it's language is not "und".  This patch results in the language being
set to "und".

writes 3gpp track names for all language variations in metadata.
when the language is not specified, uses "und" as default.

HandBrake doesn't actually allow multiple track name languages, but one
of the side effects of this patch is to use "und" when the track name
language is unspecified. Before this patch, language specific variants
of metadata were ignored and "eng" was always set as the language.

jstebbins commented:
Found some time, language should be "und" now.

galad87 commented:
It works now :)

![Schermata 2019-08-06 alle 22 52 02](https://user-images.githubusercontent.com/4305519/62576309-ef420100-b89c-11e9-9f77-792de446700f.png)

jstebbins committed:
dvd: create track named from dvd track attributes

sr55 commented:
Commit ?

jstebbins commented:
Pushed
7e5b372ee dvd: create track named from dvd track attributes
9b997b772 ffmpeg: make track names visible regarless of locale
6d43c26b1 stream: improve track name selection
57e46db5d ffmpeg: add support to read/write mp4 'titl' track tag
3999c06cb fix build error
31dca439c muxavformat: set handler_name for subtitle track names also
17e12129b muxavformat: fix writing alternative track name in mp4 hdlr atom
a1b407bc3 CLI: add subtitle name support
5defb4d6d add subtitle name to job configuration log output
d62dfce98 add subtitle track name read/write
692ac873c a few fixes for audio track name handling
f7e706149 ffmpeg: add 'name' track udta tag reading
cc33883b8 libhb: make source audio track name available to frontends

sr55 commented:
@galad87  was the MacUI updated to support?

galad87 commented:
There isn't a way to manually set the track name in the macui yet.

galad87 commented:
How is this supposed to work if the track title is "Mono" "Stereo" or "Surround"? Passing these thru would create a name that does not reflect the actual content. Should we just ignore them?

jstebbins commented:
If the original track title is something generic describing the audio format, there's not much we can do about it.  It could also be something like "AC3 5.1", etc.  We could make a list of commonly seen useless descriptions and filter them with strcasecmp, but we'll always be adding to that list I'm afraid.

Makes me think we should remove the current automatic naming of Mono, Stereo, and Surround when the description isn't set.

galad87 commented:
I agree on removing the automatic naming of Mono, Stereo, and Surround.

bradleysepos commented:
I would like to retain the automatic naming, but can live without it.

bradleysepos commented:
To clarify, I would like the option to automatically name the output tracks (not dependent on source name being empty), but I guess this can be done manually.

I haven't looked deeply into the implementation. Can track names be set in selection behavior or only the resulting track list?

jstebbins commented:
> Can track names be set in selection behavior or only the resulting track list?

@bradleysepos Only in the resulting track list right now.  

The way track naming works right now is:
- DVD source track names are generated during scan base on track attributes
- If the source format supports/has track names, they are read during scan
- Initialization of a job from a preset populates output track names from source track names
- If an audio track has no output name when encoding is started, "Mono", "Stereo", or "Surround" is assigned based on layout.  This is the step @galad87 and I would like to remove.  Right now, if the user explicitly sets the name to an empty string, they get one of these automatic names.

Automatic naming based on things like codec and layout is problematic because it interacts badly with track name passthrough when re-encoding.  The automatically generated name no longer corresponds to what the track is. If for example you originally encoded AC3 5.1 output you would get an automatic name of "Surround". Then if you re-encoded to AAC stereo, the output track name of the re-encoded audio would still be "Surround".

Our current automatic naming of tracks for DVD sources is less problematic.  This is based on attributes from the DVD that identify the track as "commentary" or "children" etc.  These attributes would apply no matter what the output codec, layout, etc is.

I don't mind giving the user a choice, but I don't think we should default to automatic naming that includes audio format characteristics. It should be an opt-in thing. If we did naming in selection behaviour I'm thinking we would need something like our output file name template. It could include the DVD attributes and things like output codec, layout etc.  You still have the problem of deciding whether to keep the existing source track name or generate a new track name.  Maybe a few checkboxes, "Track Name Passthrough", "Automatic Track Name" (with a template entry box) and "Prefer Original Track Name".  "Prefer Original Track Name" would only be available when both of the others are checked.

But including DVD attributes in a template is also problematic because the attributes may not be present.  So how would we omit that component when generating a name from the template.  The separators in the template would still be there.  The template would have to have a way to associate a separator with an attribute so that both can be removed when an attribute is not present :-\

I'm starting to think I'm over complicating this :*(

bradleysepos commented:
> I'm starting to think I'm over complicating this

Nah, it's just complicated.

I'm okay with not automatically naming empties. Some people are going to want a way to retain the old behavior, though, and in this case it actually seems somewhat reasonable.
